crates/biome_grit_parser/tests/grit_test_suite/err/incorrect_string_prefix.grit.snap

```diff
@@ -0,0 +1,60 @@
+---
+source: crates/biome_grit_parser/tests/spec_test.rs
+expression: snapshot
+---
+## Input
+```grit
+unknown"console.log()"
+
+```
+
+## AST
+
+```
+GritRoot {
+    bom_token: missing (optional),
+    version: missing (optional),
+    language: missing (optional),
+    definitions: GritDefinitionList [],
+    pattern: missing (optional),
+    definitions_continued: GritDefinitionList [],
+    eof_token: EOF@0..23 "unknown\"console.log()\"\n" [] [],
+}
+```
+
+## CST
+
+```
+0: GRIT_ROOT@0..23
+  0: (empty)
+  1: (empty)
+  2: (empty)
+  3: GRIT_DEFINITION_LIST@0..0
+  4: (empty)
+  5: GRIT_DEFINITION_LIST@0..0
+  6: EOF@0..23 "unknown\"console.log()\"\n" [] []
```

Comment:
This isn't correct, and it needs to be addressed. 

Looking at the grammar, there are **two** definition lists:
- `definitions`
- `definitions_continued`

With `pattern` in the middle. Now, `unknown"console.log()"` **must** be placed inside a `definitions` or `definitions_continued`. It's to you to decide which one, but we can't have it inside `EOF`.

This means:
- The error recovery strategy still needs some work;
- The error recovery strategy needs more testing;
- Probably, the grammar needs to be reviewed


Can you give me an example of grammar where `definitions`, `definitions_continued` and `pattern` are defined? I might be able to give you a different point of view

Reply:
Good catch, thanks! The reason why I went with `definitions`, `definitions_continued` and `pattern` is because there should be only one pattern, but it can be anywhere among the definitions. I can provide a full example this afternoon.

Reply:
Here is a Grit example with two definitions, one before and one after the pattern:

```js
// define a lines function
function lines($string) {
    return split($string, separator=`\n`)
}

// Use the my_todo function
`module.exports = $_` as $x => my_todo(target=$x, message=`Fix this\nAnd that`)

// Define a my_todo function
function my_todo($target, $message) {
   if($message <: undefined) {
       $message = "This requires manual intervention."
   },
   $lines = lines(string = $message),
   $lines <: some bubble($result) $x where {
       if ($result <: undefined) {
            $result = `// TODO: $x`
        } else {
            $result += `\n// $x`
        }
   },
   $target_lines = lines(string = $target),
   $target_lines <: some bubble($result) $x where { $result += `\n// $x` },
   return $result,
}
```

I suppose we could also just combine them all and allow arbitrary patterns where definitions are allowed. Then we only need one list, and we could use a separate check afterwards to see if we parsed only a single pattern. It might give the opportunity to provide a nicer error message, but at the cost of a little performance.

But there might be a simpler solution as well :)

Reply:
I think I can indeed improve this by treating the one pattern as a definition as well, and simply adding a single boolean inside `DefinitionList` to check if a pattern was already seen. That way we can simplify the parsing, have nice error message when there are multiple patterns, still format correctly, even when there are multiple patterns, and it shouldn't measurably affect performance at all :)

Reply:
Here's an alternative of grammar. It's going to be more a bit more generic, so it will give more responsibilities to the parser, but it would allow for more flexibility:

```
GritRoot =
    bom: 'UNICODE_BOM'?
    version: GritVersion?
    language: GritLanguageDeclaration?
    declarations: GritDeclarationList
    eof: 'EOF'

// not sure 
GritDeclarationList = AnyGritDefinitionOrPattern*

AnyGritDefinitionOrPattern =
    AnyDefinition
    | AnyGritPattern
```

This would allow for any pattern and declaration to be placed anywhere in the code. The parser, however, should ensure that only **one pattern** is defined in the source code.

What do you think?

Reply:
That's pretty much what I ended up doing :)

Reply:
This is the commit: https://github.com/biomejs/biome/pull/1998/commits/08899354480672227ca347a7bb4993970643cabe